src/jitlayers.cpp

```diff
@@ -1458,6 +1458,166 @@ namespace {
 
         orc::ExecutionSession &ES;
     };
+
+    struct DLSymOptimizer {
+        DLSymOptimizer(bool named) JL_NOTSAFEPOINT {
+            this->named = named;
+            symbols_mutex = std::make_unique<std::mutex>();
+#define INIT_RUNTIME_LIBRARY(libname, handle) \
+            do { \
+                auto libidx = (uintptr_t) libname; \
+                if (libidx >= runtime_symbols.size()) { \
+                    runtime_symbols.resize(libidx + 1); \
+                } \
+                runtime_symbols[libidx].first = handle; \
+            } while (0)
+
+            INIT_RUNTIME_LIBRARY(NULL, jl_RTLD_DEFAULT_handle);
+            INIT_RUNTIME_LIBRARY(JL_EXE_LIBNAME, jl_exe_handle);
+            INIT_RUNTIME_LIBRARY(JL_LIBJULIA_INTERNAL_DL_LIBNAME, jl_libjulia_internal_handle);
+            INIT_RUNTIME_LIBRARY(JL_LIBJULIA_DL_LIBNAME, jl_libjulia_handle);
+
+#undef INIT_RUNTIME_LIBRARY
+        }
+
+        void *lookup_symbol(void *libhandle, const char *fname) JL_NOTSAFEPOINT {
+            void *addr;
+            jl_dlsym(libhandle, fname, &addr, 0);
+            return addr;
+        }
+
+        void *lookup(const char *libname, const char *fname) JL_NOTSAFEPOINT {
+            StringRef lib(libname);
+            StringRef f(fname);
+            std::lock_guard<std::mutex> lock(*symbols_mutex);
+            auto uit = user_symbols.find(lib);
+            if (uit == user_symbols.end()) {
+                void *handle = jl_get_library_(libname, 0);
```

Comment:
Feels rather risky to trigger this here, since you are holding many locks

Reply:
I could release the symbols_mutex lock and perhaps one iteration of the context lock, but otherwise it's technically valid I think since we never cross a lock boundary.

Reply:
I guess the other problem is that `jl_get_library_` is not marked as not being a safepoint, is it actually a safepoint?